lib/session.c

```diff
@@ -160,3 +173,209 @@ bool rm_session_was_aborted() {
+
+    dedupe.info._DEST_FD =
+        rm_sys_open(dest->path, cfg->dedupe_readonly ? O_RDONLY : O_RDWR);
+
+    if(dedupe.info._DEST_FD < 0) {
+        rm_log_error_line(
+            _("dedupe: error %i: failed to open dest file.%s"),
+            errno,
+            cfg->dedupe_readonly ? "" : _("\n\t(if target is a read-only snapshot "
+                                          "then -r option is required)"));
+        rm_sys_close(source_fd);
+        return EXIT_FAILURE;
+    }
+
+    /* fsync's needed to flush extent mapping */
+    fsync(source_fd);
+    fsync(dedupe.info._DEST_FD);
+
+    int ret = 0;
+    gint64 dedupe_chunk = max_dedupe_chunk;
+    while(bytes_deduped < source_stat.st_size && !rm_session_was_aborted()) {
+        dedupe.args.dest_count = 1;
+        /* TODO: multiple destinations at same time? */
+        dedupe.args._SRC_OFFSET = bytes_deduped;
+        dedupe.info._DEST_OFFSET = bytes_deduped;
+
+        /* try to dedupe the rest of the file */
+        dedupe.args._SRC_LENGTH = MIN(dedupe_chunk, source_stat.st_size - bytes_deduped);
+
+        ret = ioctl(source_fd, _DEDUPE_IOCTL, &dedupe);
+
+        if(ret != 0) {
+            rm_log_perrorf(_("%s returned error: (%d)"), _DEDUPE_IOCTL_NAME, ret);
+            break;
+        } else if(dedupe.info.status == _DATA_DIFFERS) {
+            if(dedupe_chunk != min_dedupe_chunk) {
+                dedupe_chunk = min_dedupe_chunk;
+                rm_log_debug_line("Dropping to %lu byte chunks after %lu bytes",
+                                  dedupe_chunk, bytes_deduped);
+            } else {
+                break;
+            }
+        } else if(dedupe.info.status != 0) {
+            rm_log_error_line("%s returned status %d", _DEDUPE_IOCTL_NAME,
+                              dedupe.info.status);
+            break;
+        } else if(dedupe.info.bytes_deduped == 0) {
+            break;
+        }
+
+        bytes_deduped += dedupe.info.bytes_deduped;
+    }
+
+    if(bytes_deduped == 0) {
+        rm_log_info_line(_("Files don't match - not deduped"));
+    } else if(bytes_deduped < source_stat.st_size) {
+        rm_log_info_line(_("Only first %lu bytes deduped - files not fully identical"),
+                         bytes_deduped);
+    }
+
+    rm_sys_close(source_fd);
+    rm_sys_close(dedupe.info._DEST_FD);
+
+    if(bytes_deduped == source_stat.st_size) {
+        return EXIT_SUCCESS;
+    }
+
+#undef _DEDUPE_IOCTL_NAME
+#undef _DEDUPE_IOCTL
+#undef _DEST_FD
+#undef _SRC_OFFSET
+#undef _DEST_OFFSET
+#undef _SRC_LENGTH
+#undef _DATA_DIFFERS
+#undef _DEST_FD
+#undef _FILE_DEDUPE_RANGE
+#undef _FILE_DEDUPE_RANGE_INFO
+#undef _MIN_LINUX_SUBVERSION
+#undef MAX_DEDUPE_CHUNK
+#undef MIN_DEDUPE_CHUNK
+
+#else
+    (void)cfg;
+    rm_log_error_line(_("rmlint was not compiled with file cloning support."))
+#endif
+
+    return EXIT_FAILURE;
+}
+
+/**
+ * *********** `rmlint --is-reflink` session main ************
+ **/
+int rm_session_is_reflink_main(RmCfg *cfg) {
```

Comment:
Ok I'll have a look at this.

Although I think `rmlint.sh` should skip paranoia check if using `rmlint --dedupe` since the dedupe ioctl does its own paranoid checks.